sw/host/opentitanlib/src/transport/ti50emulator/emu.rs

```diff
@@ -211,58 +204,45 @@ impl EmulatorProcess {
             }
         };
 
-        log::info!("Waiting for sub-process start");
-        let ready_handle = thread::spawn(move || {
-            let control_socket = UnixListener::bind(socket_path).unwrap();
-            control_socket
-                .set_nonblocking(true)
-                .expect("Can't set non-blocking socket");
-            let mut buffer = [0u8; 8];
-            let mut retry = 0;
-            for stream in control_socket.incoming() {
-                match stream {
-                    Ok(mut socket) => {
-                        let len = socket.read(&mut buffer[..]).unwrap();
-                        if PATTERN[..] == buffer[0..PATTERN.len()] {
-                            log::info!("Ti50Emulator ready");
-                        }
-                        return Ok(len);
-                    }
-                    Err(err) if err.kind() == ErrorKind::WouldBlock => {
-                        log::debug!("Wait for sub-process...");
-                        std::thread::sleep(TIMEOUT);
-                        retry += 1;
-                        if retry >= MAX_RETRY {
-                            return Err(EmuError::StartFailureCause(
-                                "Spawning Ti50Emulator sub-process timeout".to_string(),
-                            ));
-                        }
-                    }
-                    Err(_err) => {
-                        return Err(EmuError::RuntimeError(
-                            "Control socket io error".to_string(),
-                        ));
-                    }
-                }
-            }
-            Err(EmuError::StartFailureCause(
-                "Waiting for sub-process failed".to_string(),
-            ))
-        });
-
         log::info!("Spawning Ti50Emulator sub-process");
         log::info!("Command: {} {:?}", exec.display(), args_list);
-        let handle = Command::new(&exec)
+        let mut handle = Command::new(&exec)
             .args(args_list)
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
             .spawn()
             .context("Could not spawn sub-process")?;
-        self.proc = Some(handle);
-        match ready_handle
-            .join()
-            .expect("Can't join control socket thread")
-        {
-            Ok(_) => Ok(()),
-            Err(err) => Err(err.into()),
+
+        log::info!("Waiting for sub-process start");
+        let child_stdout = handle.stdout.take().expect("Failed to open stdout");
+        let mut line_iter = io::BufReader::new(child_stdout).lines();
+        loop {
```

Comment:
I see that PATTERN check is blocking now. I think there is still possibility of hanging processes during startup.
If it isn't a big problem i would like solution with some timeout/watchdog.
Second what happens  if lines will be reordered, and something else appear on stdout before PATTERN.

Reply:
Reordering is already handled, as you see in line 227, if this code sees a line on the output of the sub-process, which is not the one we are looking for, then it is just logged, and the control takes another round of the outer `loop`.

I will re-work the code to use Unix poll to be able to wait for output with a timeout.

Reply:
The new code uses low-level reading, rather than BufReader, to be able to apply a timeout.

Reply:
OK.